### PR TITLE
Voting

### PR DESCRIPTION
**Pull request by jdungan**
Clean up voting:
1) Adds signals to voting to update claim status
2) Clean up claim status view and template

### CODE REVIEW

File: auctions/models.py [hunk 1]
```diff
@@ -30,15 +30,18 @@ def __unicode__(self):
...
-            url=self.url
-        ).exclude(
-            user=self.user
-        ).aggregate(
-            Sum('offer')
-        )
-
-        return other_bids['offer__sum'] >= self.ask
+        # import ipdb;ipdb.set_trace()
+        if self.ask:
```
Comment (jdungan) [inline auctions/models.py]: Had to add this to make sure $0 ask were not met.

File: auctions/models.py [hunk 2]
```diff
@@ -223,9 +223,11 @@ def update_claim_status(sender, instance, created, **kwargs):
         claim.status = "Approved"
         claim.save()
 
-    if (claim.num_rejections / float(offers_needed)) >= 0.5:
-        claim.status = "Rejected"
-        claim.save()
+
+    if offers_needed > 0:
```
Comment (jdungan) [inline auctions/models.py]: Is this needed to protect against dividing by zero votes?

File: api/views.py [hunk 3]
```diff
@@ -97,15 +97,15 @@ class ClaimAPIView(APIView):
...
 
     def get(self, request, pk, format=None):
-        claim, votes, offers = None, [], []
+        claim = None
         claim = get_object_or_404(Claim, pk=pk)
-        votes = Vote.objects.filter(claim=claim)
-        voted = votes.filter(user=request.user).count() > 0
-        offers = Bid.objects.filter(issue=claim.issue).filter(offer__gt=0)
+        try:
+            vote = Vote.objects.filter(claim=claim, user=self.request.user)[0]
```
Comment (groovecoder) [inline api/views.py]: Should be `Vote.objects.get(claim=claim, user=self.request.user)`

File: auctions/models.py (hunk 1 above)
Comment (groovecoder) [inline auctions/models.py]: Is there a test for it?

Comment (jdungan) [inline auctions/models.py]: Yes (he says proudly)

File: auctions/models.py [hunk 4]
```diff
@@ -184,22 +211,47 @@ def __unicode__(self):
...
+        claim.save()
+
+    if offers_needed > 0:
+        if (claim.num_rejections / float(offers_needed)) >= 0.5:
+            claim.status = "Rejected"
+            claim.save()
+
+
+@receiver(post_save, sender=Vote)
+def notify_approved_claim(sender, instance, created, **kwargs):
```
Comment (groovecoder) [inline auctions/models.py]: We may want to fold this into the `update_claim_status` signal handler? I'm not sure what the proper Djangonaut™ approach is here: multiple signal handlers for models, or a single signal handler.

File: auctions/models.py [hunk 5]
```diff
@@ -184,22 +211,47 @@ def __unicode__(self):
...
+    claim = instance.claim
+    votes_needed = claim.offers.count()
+
+    if claim.num_rejections == votes_needed:
+        current_site = Site.objects.get_current()
+        # TODO: make a nicer HTML email template
+        CLAIM_APPROVED_EMAIL_STRING = """
```
Comment (groovecoder) [inline auctions/models.py]: Need to rename this `CLAIM_REJECTED_EMAIL_STRING`

File: auctions/tests/model_tests.py [hunk 6]
```diff
@@ -42,12 +42,24 @@ def _add_url(self, string):
...
+
+
+class BidTest(MarketTestCase):
     def test_ask_met(self):
         self.assertFalse(self.bid1.ask_met())
         mommy.make(Bid, ask=0, offer=10, url=self.url)
         self.assertTrue(self.bid1.ask_met())
 
+    def test_zero_ask_return_false(self):
+        self.assertEqual(self.bid3.ask_met(), False)
```
Comment (groovecoder) [inline auctions/tests/model_tests.py]: There's the test! :smile:

File: auctions/tests/model_tests.py [hunk 7]
```diff
@@ -290,6 +349,13 @@ def setUp(self):
         mommy.make(Bid, user=self.user3, url=url, issue=issue, offer=50)
         self.claim = mommy.make(Claim, issue=issue, user=self.user1)
 
+    def test_unicode(self):
```
Comment (groovecoder) [inline auctions/tests/model_tests.py]: heh, I love these `test_unicode` tests trying to get to 100% coverage. :wink:

Comment (jdungan) [inline auctions/tests/model_tests.py]: You've created a monster.  I'm very troubled that our coverage is below 97%.

**Comment by groovecoder** [comment]
This is looking good. Still some gnarly `YOU` and `SOMEONE` template syntax and conditionals, but we can sweep thru them all again another time.

**Comment by groovecoder** [comment]
I'm planning to spot-check and merge this and then hit the next one that's built on top of this.

**Comment by jdungan** [comment]
Slogging through the payments stuff right now.I've caught some of the easy changes in my current work as you made the comments.

**Comment by groovecoder** [comment]
Works!!

File: api/views.py (hunk 3 above)
Comment (jdungan) [inline api/views.py]: 60877a6

File: auctions/models.py (hunk 5 above)
Comment (jdungan) [inline auctions/models.py]: https://github.com/codesy/codesy/commit/74a49f0b7d22a4d672e0644a0648e0430242eb10

**Pull request merged**